notes/lsp-semantics.md

```diff
@@ -0,0 +1,146 @@
+
+### Proposed algorithm
+
+The core part of the algorithm is a function mapping terms to sets of
+record literals, giving for each term the set of record literals that might
+"contribute" (through, e.g., merging) to that term. This will only be an
+approximation, of course, because we aren't evaluating. Let's call this process
+"record literal resolution," and it is defined by:
+
+- a record literal resolves to itself
+- a variable resolves to the term that it's bound to
+- a merge resolves to the union of the resolutions of the two merged terms
+- an if-then-else resolves to the union of the resolutions of the two branches
+- everything else resolves to the empty set (but this can be extended later)
+
+Now to find the definition of `baz` in `foo.bar.baz`, we first resolve `foo`
+to a set of record literals. For each of those literals containing a `bar` field,
+we resolve their values to a set of record literals. For each of those literals
+containing a `baz` field, we report it as a definition of `baz`.
+
+## Goto type definition
+
+Since we cannot (yet) name custom types, it probably doesn't make sense to use "goto type definition"
+for static types. Instead, we could use the LSP "goto type definition" request for going
+to contract definitions: given `x | Foo`, a "goto type definition" request for `x` should be the same
+as a "goto definition" request for `Foo`. This would respect record contracts, so that
+in `{x = 1, y = 2} | { x | Foo, y | Bar }` a "goto type definition" request for `x` should again
+be the same as a "goto definition" request for `Foo`.
+
+## Completion
+
+Here is a list of items that we might want to provide completions for:
+
+- record fields, as in `let x = { foo = 1 } in x.fo`
+- enum variants
+- variables in scope, as in `let foo = 1 in 2 + fo`
+- filenames in imports, as in `import "fo` when `foo.ncl` exists on disk
+- maybe keywords? They're pretty short in nickel
+
+One of the trickier parts of handling completion is that the input will be incomplete
+and may not parse. Let's ignore that for now, and assume that we have a full AST.
+
+In LSP, the editor (and not the language server) is in charge of text-based filtering.
+That is, if the user enters `2 + fo` and requests completion, we can return all of the
+names of variables in scope; the editor is in charge of filtering out all those that
+don't start with "fo". When types are involved, the responsibilities swap: when
+completing `x.fo` the language server should return only the field names belonging
+to `x` (it still doesn't need to care about whether they start with "fo").
+Type information can also be used to filter completions for enum variants and
+variables in scope (more on that below).
+
+In summary, the desired completion behavior is:
+
+- when completing a record path, find the type (or contract) of the second-last
+  element of the path. If it's a record type, return all of its fields.
```

Comment:
I think this part is crucial and is already currently more involved. First, I believe"find the type" isn't always that simple. Maybe it would deserve a good operational definition?

Additionally, most of the time, the type isn't a record type exactly (might be a contract for example). What about something like `{foo | Contract1 & Contract2}`, where `Contract1` and `Contract2` are records? Or even just when `{ foo | Contract}` is a record contract. In all generality, you even have to handle `foo.bar.baz`, thus need to define how to extract this information from a top-level contract.

Taking a step back, I believe the work we need to do for completion and for resolution is not that different. In essence, you want to associate to any expression (most of the time, a variable, but we might extend later) a set `fields_infos: HashMap<Ident, DefInfo>` where `DefInfo` could be something like:

```nickel
struct Definfo {
  definitions: HashMap<Ident, DefInfo>, // what are the definitions available on this field
  body: Option<TermPos>, // the location of the definition, if any, for goto definition
}
```

This structure can use all the information available: type annotations, contract annotations, and the structure of the term. So, in essence, you want to define `field_infos(RichTerm) -> DefInfo` in pseudo code, something like:

```
field_infos(e1 & e2) = field_infos(e1) U field_infos(e2)
field_infos(let x = e1 in e2) = field_infos(e2)
field_infos(fun x => body) = field_infos(body)
field_infos(head x) = field_infos(head)
field_infos(var) = field_infos(goto_definition(var))
field_infos(e1 | C) = field_infos(e1) U extract_field_infos(C)
field_infos(foo.bar) = field_infos(goto_definition(bar in foo.bar))
other cases => empty
```

I just wrote that arbitrarily, so some rules might not be entirely what we want, and I probably forgot important constructions, but you get it. Then it'll be quite easy to implement the corresponding function, following the specification.

Then, you can define the completion of `foo.bar.baz` as `field_infos(foo.bar).definitions.keys().collect()` or something.

I believe you could even fit the completion of variable in this story, seeing it as completing the "record" corresponding to the environment at the location of completion (which we already record in the linearization).